cro-clib/src/wallet.rs

```diff
@@ -0,0 +1,233 @@
+        match src {
+            ApiNetwork::Mainnet => Network::Mainnet,
+            ApiNetwork::Testnet => Network::Testnet,
+            ApiNetwork::Devnet => Network::Devnet,
+        }
+    }
+}
+
+#[allow(dead_code)]
+unsafe fn get_string(src: *const c_char) -> String {
+    CStr::from_ptr(src).to_string_lossy().into_owned()
+}
+
+#[allow(dead_code)]
+fn copy_string(src: &str, dst: &mut [u8]) {
+    dst[..src.len()].copy_from_slice(&src.as_bytes()[..src.len()]);
+    dst[src.len()] = 0;
+}
+
+/// hdwallet creating using bip44 hdwallet
+#[no_mangle]
+pub unsafe extern "C" fn cro_create_hdwallet(
+    wallet_out: *mut HDWalletPtr,
+    mnemonics: *mut u8,
+    mnemonics_length: i32,
+) -> i32 {
+    let mnemonic = Mnemonic::new();
+    let phrase = mnemonic.unsecure_phrase();
+    assert!(phrase.as_bytes().len() < mnemonics_length as usize);
+    let wallet = HDWallet {
+        seed: HDSeed::from(&mnemonic),
+    };
+    let wallet_box = Box::new(wallet);
+    ptr::write(wallet_out, Box::into_raw(wallet_box));
+    ptr::copy_nonoverlapping(
+        phrase.as_bytes().as_ptr(),
+        mnemonics,
+        phrase.as_bytes().len(),
+    );
+    SUCCESS
+}
+
+/// restore bip44 hdwallet from mnemonics which user gives
+#[no_mangle]
+pub unsafe extern "C" fn cro_restore_hdwallet(
+    mnemonics_string: *const c_char,
+    wallet_out: *mut HDWalletPtr,
+) -> i32 {
+    let mnemonics = get_string(mnemonics_string);
+    let mnemonics_sec = SecUtf8::from(mnemonics);
+    let mnemonic = Mnemonic::from_secstr(&mnemonics_sec).unwrap();
+    let wallet = HDWallet {
+        seed: HDSeed::from(&mnemonic),
+    };
+    let wallet_box = Box::new(wallet);
+    ptr::write(wallet_out, Box::into_raw(wallet_box));
+    SUCCESS
+}
+
+/// create staking address from bip44 hdwallet
+#[no_mangle]
+pub unsafe extern "C" fn cro_create_staking_address(
+    wallet_ptr: HDWalletPtr,
+    network: ApiNetwork,
+    address_out: *mut AddressPtr,
+    index: u32,
+) -> i32 {
+    if wallet_ptr.is_null() {
+        return FAIL;
+    }
+    let wallet = wallet_ptr.as_mut().expect("get wallet");
+    let (public, private) = wallet
+        .seed
+        .derive_key_pair(network.into(), Account::Staking as u32, index)
+        .unwrap();
+    let address = StakedStateAddress::BasicRedeem(RedeemAddress::from(&public));
+
+    let ret = Address {
+        privatekey: private,
+        publickey: public,
+        address: address.to_string(),
+    };
+    let address_box = Box::new(ret);
+    ptr::write(address_out, Box::into_raw(address_box));
+    SUCCESS
+}
+
+/// print address information
+#[no_mangle]
+pub unsafe extern "C" fn cro_print_address(address_ptr: AddressPtr) -> i32 {
+    let address = address_ptr.as_mut().expect("get address");
+    println!("{}", address.address.to_string());
+    SUCCESS
+}
+
+/// print address information
+#[no_mangle]
+pub unsafe extern "C" fn cro_get_address(
+    address_ptr: AddressPtr,
+    buf: *mut u8,
+    buf_length: i32,
```

Comment:
buf is a bit generic name -- perhaps "address_output" ? + length type can be usize/u32 instead of i32

Reply:
ok